Data/ContentStore.hs

```diff
@@ -133,11 +172,49 @@ doStore cs hasher writer object = do
 
     -- Only store the object if it does not already exist in the content store.
     -- If it's already there, just return the digest.
-    unlessM (liftIO $ doesFileExist path) $
-        liftIO $ writer path object
+    unlessM (liftIO $ doesFileExist path) $ liftIO $ do
+        (tmpPath, handle) <- startStore cs
+        writer handle object
+        finishStore cs (tmpPath, handle) digest
```

Comment:
I don't know that it means you should do anything, but this feels an awful lot like `bracket`.

Reply:
The thing that makes it not like bracket is that `finishStore` isn't a finalizer. It should only run if the writer succeeds. Maybe there's some kind of Applicative operator tricks I could do to make the sequencing more obvious but I'm not real sure